modules/apps/message-boards/message-boards-service/src/main/java/com/liferay/message/boards/internal/util/MBMessageUtil.java

```diff
@@ -17,19 +17,175 @@
+			_getMarkupElement(MarkupElement.END, htmlFormat),
+			_getMarkupElement(MarkupElement.START_BODY_ROOT, htmlFormat),
+			getMessageBody(htmlFormat, mbMessage, quoteMark, serviceContext),
+			_getMarkupElement(MarkupElement.END, htmlFormat),
+			_getMarkupElement(MarkupElement.END, htmlFormat));
+	}
+
+	public static String getSameLevelMessages(
+		boolean htmlFormat, List<MBMessage> mbMessages, long messageId,
+		String quoteMark, ServiceContext serviceContext) {
+
+		if (ListUtil.isEmpty(mbMessages)) {
+			return StringPool.BLANK;
+		}
+
+		StringBuilder messageSiblings = new StringBuilder(
+			mbMessages.size() + 2);
+
+		messageSiblings.append(
+			_getMarkupElement(MarkupElement.START_SIBLING, htmlFormat));
+
+		for (MBMessage childMessage : mbMessages) {
+			if (childMessage.getMessageId() != messageId) {
+				messageSiblings.append(
+					StringBundler.concat(
+						_getMarkupElement(
+							MarkupElement.START_MESSAGE_SIBLING, htmlFormat),
+						_getMarkupElement(
+							MarkupElement.START_USER_SIBLING, htmlFormat),
+						_getUserName(htmlFormat, childMessage, quoteMark),
+						_getMarkupElement(MarkupElement.END, htmlFormat),
+						_getMarkupElement(
+							MarkupElement.START_BODY_SIBLING, htmlFormat),
+						getMessageBody(
+							htmlFormat, childMessage, quoteMark,
+							serviceContext),
+						_getMarkupElement(MarkupElement.END, htmlFormat),
+						_getMarkupElement(MarkupElement.END, htmlFormat)));
+			}
+		}
+
+		messageSiblings.append(
+			_getMarkupElement(MarkupElement.END, htmlFormat));
+
+		return messageSiblings.toString();
+	}
+
+	public static String getThreadMessages(
+		Boolean htmlFormat, String quoteMark, List<MBMessage> thread,
+		ServiceContext serviceContext) {
+
+		if (ListUtil.isEmpty(thread)) {
+			return StringPool.BLANK;
+		}
+
+		StringBuilder parentMessageBody = new StringBuilder(thread.size());
+		StringBuilder endElements = new StringBuilder(thread.size());
+
+		for (int i = 0; i < thread.size(); i++) {
+			MBMessage mbMessage = thread.get(i);
+
+			parentMessageBody.append(
+				StringBundler.concat(
+					_getMarkupElement(MarkupElement.START_MESSAGE, htmlFormat),
+					_getMarkupElement(
+						MarkupElement.START_USER_MESSAGE, htmlFormat),
+					_getUserName(
+						htmlFormat, mbMessage, quoteLeveled(i, quoteMark)),
+					_getMarkupElement(MarkupElement.END, htmlFormat),
+					_getMarkupElement(
+						MarkupElement.START_MESSAGE_BODY, htmlFormat),
+					getMessageBody(
+						htmlFormat, mbMessage, quoteLeveled(i, quoteMark),
+						serviceContext),
+					_getMarkupElement(MarkupElement.END, htmlFormat),
+					_getMarkupElement(MarkupElement.END_MESSAGE, htmlFormat)));
+
+			endElements.append(
+				_getMarkupElement(MarkupElement.END_ELEMENT, htmlFormat));
+		}
+
+		return StringBundler.concat(
+			_getMarkupElement(MarkupElement.START_MESSAGE_THREAD, htmlFormat),
+			parentMessageBody.toString(), endElements.toString(),
+			_getMarkupElement(MarkupElement.END, htmlFormat));
+	}
```

Comment:
There's no need to have two string bundlers. Simply keep a count of the number of end elements that you need, and when the loop ends, do something like:
```
// Add this at the beginning, not at the end.
sb.append(_getMarkupElement(MarkupElement.START_MESSAGE_THREAD, htmlFormat);

for (int i = 0; i < thread.size(); i++) {
    ...
    endElementCount++;
    ...
}

for (int i = 0; i < endElementCount; i++) {
    sb.append(_getMarkupElement(MarkupElement.END_ELEMENT, htmlFormat));
}

sb.append(_getMarkupElement(MarkupElement.END, htmlFormat));

return sb.toString();
```

Reply:
Also, don't call a list of MBMessage "thread"; it is not an mbThread, so you should name it "mbMessages" (or "messages", but keep the naming consistent in the class).